src/util/mapbox.js

```diff
@@ -119,3 +126,108 @@ function formatUrl(obj: UrlObject): string {
+        this.queue = [];
+        this.pending = false;
+        this.pendingRequest = null;
+    }
+
+    queueRequest(date: number) {
+        this.queue.push(date);
+        this.processRequests();
+    }
+
+    processRequests() {
+        if (this.pending || this.queue.length === 0) {
+            return;
+        }
+        this.pending = true;
+        let dueForEvent = false;
+        if (!this.eventData.anonId || !this.eventData.lastSuccess) {
+            //Retrieve cached data
+            if (this.localStorageAvailable) {
+                try {
+                    const data = window.localStorage.getItem(this.STORAGE_TOKEN);
+                    if (data) {
+                        const json = JSON.parse(data);
+                        this.eventData.anonId = json.anonId;
+                        this.eventData.lastSuccess = Number(json.lastSuccess);
+                    }
+                } catch (e) {
+                    warnOnce('Unable to read from LocalStorage');
+                }
+            }
+        }
+
+        if (!validateUuid(this.eventData.anonId)) {
+            this.eventData.anonId = uuid();
+            dueForEvent = true;
+        }
+        const nextUpdate = this.queue.shift();
+
+        // Record turnstile event once per calendar day.
+        if (this.eventData.lastSuccess) {
+            const lastUpdate = new Date(this.eventData.lastSuccess);
+            const nextDate = new Date(nextUpdate);
+            const daysElapsed = (nextUpdate - this.eventData.lastSuccess) / (24 * 60 * 60 * 1000);
+            dueForEvent = dueForEvent || daysElapsed >= 1 || daysElapsed < 0 || lastUpdate.getDate() !== nextDate.getDate();
+        }
+        if (!dueForEvent) {
+            return;
+        }
+
+        const evenstUrlObject: UrlObject = parseUrl(config.EVENTS_URL);
+        evenstUrlObject.params.push(`access_token=${config.ACCESS_TOKEN || ''}`);
+        const request: RequestParameters = {
+            url: formatUrl(evenstUrlObject),
+            headers: {
+                'Content-Type': 'text/plain' //Skip the pre-flight OPTIONS request
+            }
+        };
+
+        const payload = JSON.stringify([{
+            event: 'appUserTurnstile',
+            created: (new Date(nextUpdate)).toISOString(),
+            sdkIdentifier: 'mapbox-gl-js',
+            sdkVersion: `${version}`,
+            'enabled.telemetry': false,
+            userId: this.eventData.anonId
+        }]);
+
+        this.pendingRequest = postData(request, payload, (error) => {
+            if (!error && this.localStorageAvailable) {
+                this.eventData.lastSuccess = nextUpdate;
+                window.localStorage.setItem(this.STORAGE_TOKEN, JSON.stringify({
+                    lastSuccess: this.eventData.lastSuccess,
+                    anonId: this.eventData.anonId
+                }));
+                this.pendingRequest = null;
+                this.pending = false;
+                this.processRequests();
```

Comment:
Definitely a bug that the `pending..` variables aren't cleared. For `processRequests` I think it should stop attempting to send requests if there is an error, until the next event is queued - in case of network connection loss, or a server outage.

Reply:
Ah, yeah that seems reasonable